src/main/java/io/vertx/redis/client/Redis.java

```diff
@@ -59,24 +61,37 @@ static Redis createClient(Vertx vertx, String connectionString) {
    * @return the client
    */
   static Redis createClient(Vertx vertx, RedisOptions options) {
+    return createClient(vertx, options, () -> Future.succeededFuture(new RedisOptions(options)));
+  }
+
+  /**
+   * Create a new redis client using the given client options.
+   *
+   * @param vertx the vertx instance
+   * @param options the user provided options
+   * @param optionsSupplier the user provided options, which can be changed dynamically.
+   *                        Values like type cannot be changed.
+   * @return the client
+   */
+  @GenIgnore
+  static Redis createClient(Vertx vertx, RedisOptions options, Supplier<Future<RedisOptions>> optionsSupplier) {
```

Comment:
I think here we should only use the supplier and avoid the extra option. We will simply use the first option returned by the supplier to have the fixed data. That is what we did in vertx-sql-client.

After in master we should refactor the options to split between the per connection option and the general options.

Reply:
Ok. Two questions:
1. In the create method the type is used to determine, which client to create. This would force me to resolve the future directly in createClient, which would block. I am not that familiar with the vertx future. The only method I could find is ```result()```, but when I check the implementation, then value might be null. So the only other way I could see is: ```optionsSupplier.get().toCompletionStage().toCompletableFuture().get()```, which looks strange :)
2. Do you want that I also remove the ```options``` from all the client constructor? Or use the resolved options mentioned from question 1?

Reply:
1. why do we need an extra options that we need to resolve ?
2. we need to keep backward compatibility, so the simple options, should be wrapped as a supplier that always returns the same options

Reply:
Hi @vietj . I think I did not describe my issue properly. See the potential change below. I though with you first comment you want me to remove the additional RedisOptions. When I do that I only have the Supplier<Future<>>, which I need to resolve immediately. Otherwise, I have no "type" to check against. And in the example below I do not know how to resolve a vertx future in a non-reactive block.  Therefore, the approach I mentioned above.

```java
  /**
   * Create a new redis client using the given client options.
   * @param vertx the vertx instance
   * @param options the user provided options
   * @return the client
   */
  static Redis createClient(Vertx vertx, RedisOptions options) {
    return createClient(vertx, () -> Future.succeededFuture(new RedisOptions(options)));
  }

  /**
   * Create a new redis client using the given client options.
   *
   * @param vertx the vertx instance
   * @param optionsSupplier the user provided options, which can be changed dynamically.
   *                        Values like type cannot be changed.
   * @return the client
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static Redis createClient(Vertx vertx, Supplier<Future<RedisOptions>> optionsSupplier) {
    RedisOptions options;
    try {
      options = optionsSupplier.get().toCompletionStage().toCompletableFuture().get();
    } catch (InterruptedException | ExecutionException e) {
      throw new IllegalStateException("Could not resolve initial RedisOptions.");
    }

    switch (options.getType()) {
      case STANDALONE:
        return new RedisClient(vertx, options, optionsSupplier);
      case SENTINEL:
        return new RedisSentinelClient(vertx, options, optionsSupplier);
      case CLUSTER:
        return new RedisClusterClient(vertx, options, optionsSupplier);
      case REPLICATION:
        return new RedisReplicationClient(vertx, options, optionsSupplier);
      default:
        throw new IllegalStateException("Unknown Redis Client type: " + options.getType());
    }
  }
```

The second question is related to the switch block. There I need to create the different clients based on the "type". E.g.
```java
return new RedisClient(vertx, options, optionsSupplier);
```

My question here was if I should also remove the options parameter and only keep optionsSupplier.

Reply:
ok that makes sense to me, we need to find a way to make this work.

There are several ways to solve it, I believe the easiest way is to have the createClient method returns a future instead of the type.

We could also think instead using a lazy RedisClient interface which has a volatile `Future<RedisClient>` field that is initialised when first accessed.

Reply:
Actually I am thinking that this option should only be possible for pooled connections and not for direct connections (for which it does not make sense).

Reply:
So my point is that the type specificity RedisClient/RedisSentinelClient/RedisClusterClient/... should be moved more internally which I think requires a refactor of the client for this.

We should only have internally a single client class and the subclasses of `Redis` should be refactored as a connection factory hierarchy like we did in the Vert.x SQL client.

Reply:
maybe you can just provide the PR as is and I would take care of the refactoring if that is too much work for you.

Reply:
Hi @vietj ,
that would be awesome actually, because I would struggle and annoy you all the time to make sure I am not messing it up. 
Thank you very much,
Chris